core/store/src/flat_state.rs

```diff
@@ -209,26 +240,81 @@ mod imp {
 pub use imp::{FlatState, FlatStateFactory};
-use near_primitives::hash::CryptoHash;
-use near_primitives::types::BlockHeight;
+use near_primitives::state::ValueRef;
+use near_primitives::types::{BlockHeight, RawStateChangesWithTrieKey, ShardId};
 use std::collections::HashMap;
-use std::sync::{Arc, RwLock};
 
-#[allow(unused)]
-pub struct FlatStateDelta {
-    // TODO: add delta implementation
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct KeyForFlatStateDelta {
+    pub shard_id: ShardId,
+    pub block_hash: CryptoHash,
 }
 
+/// Delta of the state for some shard and block, stores mapping from keys to value refs or None, if key was removed in
+/// this block.
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct FlatStateDelta(pub HashMap<Vec<u8>, Option<ValueRef>>);
+
 impl FlatStateDelta {
-    #[allow(unused)]
-    pub fn from_wrapped_trie_changes(trie_changes: &WrappedTrieChanges) -> Self {
-        // TODO: implement
-        Self {}
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    /// Creates delta using raw state changes for some block.
+    pub fn from_state_changes(changes: &[RawStateChangesWithTrieKey]) -> Self {
+        let mut delta = HashMap::new();
+        for change in changes.iter() {
+            let key = change.trie_key.to_vec();
+            if near_primitives::state_record::is_delayed_receipt_key(&key) {
+                continue;
+            }
+
+            // `RawStateChangesWithTrieKey` stores all sequential changes for a key within a chunk, so it is sufficient
+            // to take only the last change.
+            let last_change = &change
+                .changes
+                .last()
+                .expect("Committed entry should have at least one change")
+                .data;
+            match last_change {
+                Some(value) => {
+                    delta.insert(key, Some(near_primitives::state::ValueRef::new(value)))
+                }
+                None => delta.insert(key, None),
+            };
+        }
+        Self(delta)
     }
+
+    /// Applies delta to the flat state.
+    #[cfg(feature = "protocol_feature_flat_state")]
+    pub fn apply_to_flat_state(self, store_update: &mut StoreUpdate) {
+        for (key, value) in self.0.into_iter() {
+            match value {
+                Some(value) => {
+                    store_update
+                        .set_ser(crate::DBCol::FlatState, &key, &value)
+                        .expect("Borsh cannot fail");
+                }
+                None => {
+                    store_update.delete(crate::DBCol::FlatState, &key);
+                }
+            }
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_flat_state"))]
+    pub fn apply_to_flat_state(self, _store_update: &mut StoreUpdate) {}
 }
 
+#[cfg(feature = "protocol_feature_flat_state")]
+use near_primitives::block_header::BlockHeader;
+use std::sync::{Arc, RwLock};
+
 pub enum FlatStateDeltaDir {
```

Comment:
Let's remove FlatStateDeltaDir since we are not going to implement the backwards direction for now.